### PR TITLE
Frontend Chat Feature Integration Part 1

### PR DESCRIPTION
**Pull request by Ememobong28**
**Description**
This pull request focuses on integrating the chat feature into our frontend.
- Added the chat component to the frontend, allowing users to interact with each other through chat messages.
- Ensured that the chat feature is accessible only to authenticated users by checking the current user's authentication status.
- Fetched and displayed the list of contacts, allowing users to initiate chats with other users.
- Improved the appearance of chat messages, with sender and receiver names correctly aligned and timestamps displayed on a new line after each message.
- Introduced a message search functionality, enabling users to find specific messages within the chat history.

**Milestones:**
This PR contributes to the following features:
- Seamless Chat Integration: Users can now access the chat feature and interact with each other.
- Enhanced Functionality: The addition of message search improves communication and usability.

**Test Plan:**
- Conducted manual testing to verify that authenticated users can access the chat feature.
- Checked that the list of contacts is correctly displayed, and users can start conversations with other users.
- Tested the message search functionality to ensure it accurately displays matching messages.

Screenshot 2023-07-30 at 11 42 40 PM

### CODE REVIEW

File: techhive-ui/src/components/Chat/Chat.jsx [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        // }
+      }
+    };
+  
+    fetchContacts();
+  }, [currentUser, navigate]);
+  
+  console.log(allUsersRoute)
+  
+  const handleChatChange = (chat) => {
```
Comment (nicolec2626) [inline techhive-ui/src/components/Chat/Chat.jsx:57]: this might be better named onChatClicked

File: techhive-ui/src/components/Chat/Chat.jsx [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            const response = await axios.get(`${allUsersRoute}`);
+            const data = response.data;
+            console.log("Contacts API Response:", data); // Log the API response
+            setContacts(data);
+          } catch (error) {
+            console.error("Error fetching contacts:", error);
+          }
+        // } else {
+        //   // navigate("/setAvatar");
+        // }
```
Comment (nicolec2626) [inline techhive-ui/src/components/Chat/Chat.jsx:48]: we don't want to commit commented out code

**Comment by Ememobong28** [comment]
**Enhancement to Chat Feature: Adding Timestamp, Date Selector and Search Functionality**

**Description:**
In this update, We Introduced three new features: a timestamp for each message, a date selector for easy navigation through message history, and a search functionality for efficient retrieval of specific messages.

1.)**Timestamp Feature**: With this feature, each message in the chat will be tagged with the exact date and time it was sent or received, thus providing a precise context of communication.

2.) **Date Selector Feature**: The date selector will facilitate easy and fast navigation through past messages. Users can select a specific date and see all the conversations that happened on that day instantly.

3.) **Search Functionality**: A search bar will be added to the chat interface. Users can type in keywords or phrases, and the system will return all messages that contain the entered search query. This feature will greatly improve the ability to locate specific messages within the chat history.

File: techhive-ui/src/components/ChatContainer.jsx [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+            .filter((msg) => msg.sender === user.username)
+            .map((msg) => ({
+              ...msg,
+              timestamp: parseISOString(msg.createdAt),
+              originalTimestamp: msg.createdAt,
+            }));
+      
+          // Filter messages where the sender is the current chat user (received messages)
+          const receivedMessages = messagesFromServer
+            .filter((msg) => msg.sender === currentChat.username)
```
Comment (nicolec2626) [inline techhive-ui/src/components/ChatContainer.jsx:38]: what's the difference between currentChat.username and user.username above? 

maybe the naming can be improved here

File: techhive-ui/src/components/ChatContainer.jsx [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+      };
+
+  if (currentChat && user) {
+    fetchMessages();
+  }
+}, [currentChat, user]);
+
+  
+ const handleSendMsg = async (msg) => {
+      try {
```
Comment (nicolec2626) [inline techhive-ui/src/components/ChatContainer.jsx:59]: nit: I think you have a bit of formatting issue here

File: techhive/controllers/messageController.js [hunk 5]
```diff
@@ -5,6 +5,8 @@ import mongoose from 'mongoose';
 export async function getMessages(req, res, next) {
     try {
       const { from, to } = req.body;
+
+      console.log("req.body: ",req.body)
```
Comment (nicolec2626) [inline techhive/controllers/messageController.js:9]: again, we typically use console.logs just for debugging purposes

**Pull request merged**